picard/metadata.py

```diff
@@ -244,6 +256,35 @@ def compare_to_release_parts(self, release, weights):
         except (ValueError, KeyError):
             pass
 
+        # Date Logic
+        date_match_factor = 0.0
+        if "date" in release and release['date'] != '':
+            release_date = release['date']
+            release_year = extract_year_from_date(release_date)
+            if "date" in self:
+                metadata_date = self['date']
+                # release has a date and it matches what our metadata had exactly.
+                if release_date == metadata_date:
+                    date_match_factor = self.__date_match_factors['exact']
+                else:
+                    metadata_year = extract_year_from_date(metadata_date)
+                    # release has a date and it matches what our metadata had for year exactly.
+                    if release_year == metadata_year:
+                        date_match_factor = self.__date_match_factors['year']
+                    # release has a date and it matches what our metadata had closely (year +/- 2).
+                    elif abs(release_year - metadata_year) <= 2:
+                        date_match_factor = self.__date_match_factors['close_year']
+                    # release has a date but it does not match ours(all else equal, its better to have an unknown date than a wrong date. since the unknown could actually be correct)
+                    else:
+                        date_match_factor = self.__date_match_factors['differed']
+            # release has a date but we don't have one (all else equal, we prefer tracks that have non-blank date values)
+            else:
+                date_match_factor = self.__date_match_factors['exists_vs_null']
+        # release has a no date (all else equal, we don't prefer this release since its date is missing)
+        else:
```

Comment:
I just noticed you placed comments outside blocks they belong to.

I think  it is much better to keep the comment in the code block (and wrap long comments):

```python
       # Date Logic
        date_match_factor = 0.0
        if "date" in release and release['date'] != '':
            release_date = release['date']
            release_year = extract_year_from_date(release_date)
            if "date" in self:
                metadata_date = self['date']
                if release_date == metadata_date:
                    # release has a date and it matches what our metadata had exactly.
                    date_match_factor = self.__date_match_factors['exact']
                else:
                    metadata_year = extract_year_from_date(metadata_date)
                    if release_year == metadata_year:
                        # release has a date and it matches what our metadata had for year exactly.
                        date_match_factor = self.__date_match_factors['year']
                    elif abs(release_year - metadata_year) <= 2:
                        # release has a date and it matches what our metadata had closely (year +/- 2).
                        date_match_factor = self.__date_match_factors['close_year']
                    else:
                        # release has a date but it does not match ours (all else equal,
                        # its better to have an unknown date than a wrong date, since
                        # the unknown could actually be correct)
                        date_match_factor = self.__date_match_factors['differed']
            else:
                # release has a date but we don't have one (all else equal, we prefer
                # tracks that have non-blank date values)
                date_match_factor = self.__date_match_factors['exists_vs_null']
        else:
            # release has a no date (all else equal, we don't prefer this
            # release since its date is missing)
            date_match_factor = self.__date_match_factors['no_release_date']
```

@samj1912 what do you think?

Reply:
done and committed.  thanks for typing it out to show me what you meant.

Reply:
Thanks.

Perhaps you can just squash last 4 commits

Reply:
sure, I did the squash